### PR TITLE
Add Cardinal to SALAMANDER

### LINKED ISSUE #38: Add Cardinal to SALAMANDER
## Reason
[Cardinal](https://github.com/neams-th-coe/cardinal) enables coupling of OpenMC to SALAMANDER in order to perform neutronics simulation for fusion applications, providing a heating rate to the SALAMANDER heat conduction models and a production rate to the tritium transport simulation from TMAP8. It needs to be added to SALAMANDER as a submodule and then properly linked/registered. 

## Design

1. Add Cardinal as a SALAMANDER submodule.
2. Modify the SALAMANDER `Makefile` in order to properly detect and link Cardinal.
3. Modify `FenixApp.C` in order to register Cardinal objects.
4. Update SALAMANDER getting started instructions given potential Cardinal setup procedure additions.

## Impact
Adds new capability to SALAMANDER and enables completion of #30.

### PR DESCRIPTION
**Pull request by cticenhour**
Closes #38 

@meltawila As discussed earlier, please build this version of SALAMANDER and run it against your Cardinal inputs from #30 (and any other inputs you are working on, as you like). Please also add any comments you have on build instructions and usage issues to this thread so that they can be addressed as part of the review process. Thanks! 

Requesting @simopier for the formal review once this test run is complete and any comments from that test are fully addressed.

### CODE REVIEW

File: doc/content/getting_started/installation.md
Comment (cticenhour) [inline doc/content/getting_started/installation.md:90]: Reminder to add a warning to the documentation here letting the user know that this must be set in the environment *every time* the Terminal is reloaded (unless it is set persistently in the shell profile).

**Comment by cticenhour** [comment]
Note that some changes to the recipes need to occur before this can be merged. I am working on a recipe PR that I will get reviewed and merged next week.

**Comment by cticenhour** [comment]
All - also happy to take feedback on the script configuration currently used here. An alternative is doing everything in the Makefile, which should remove at least one step from the process as currently outlined, and require fewer recipe changes.

File: scripts/update_and_rebuild_cardinal.sh
Comment (meltawila) [inline scripts/update_and_rebuild_cardinal.sh]: I think this is never set automatically via conda (at least on my system). `export HDF5_ROOT=$CONDA_PREFIX` works for me

**Comment by meltawila** [comment]
Thank you Casey! I was able to follow your instructions and install with Cardinal enabled. I ran Cardinal test suite as well and got 384 passed, 254 skipped, 0 pending, 6 FAILED. The failed tests are probably out of the scope of this PR (i.e. one of them uses THM which we're not including here). otherwise it looks great to me!

File: scripts/update_and_rebuild_cardinal.sh
Comment (meltawila) [reply scripts/update_and_rebuild_cardinal.sh]: I also used `export LD_LIBRARY_PATH=$CONDA_PREFIX/lib:$LD_LIBRARY_PATH` which I know cause errors sometimes when it's not set properly

File: doc/content/getting_started/installation.md
Comment (simopier) [reply doc/content/getting_started/installation.md:90]: > Reminder to add a warning to the documentation here letting the user know that this must be set in the environment every time the Terminal is reloaded (unless it is set persistently in the shell profile).

Are you going to add this reminder in this PR?

File: doc/content/getting_started/installation.md
Comment (simopier) [inline doc/content/getting_started/installation.md]:
```suggestion
cd ~/projects/FENIX
scripts/update_and_rebuild_cardinal.sh
```

**Review by simopier — CHANGES_REQUESTED** [review]
Looks good to me. Thank you @cticenhour for adding this!
I'll re-review after the new recipe is added.

**Comment by simopier** [comment]
Converting as draft until we get the new recipe merged.

**Comment by simopier** [comment]
I think it would be nice to get @loganharbour's review as well.

File: doc/content/getting_started/installation.md
Comment (cticenhour) [reply doc/content/getting_started/installation.md:90]: Yes - I like to avoid too many changes after I've requested reviews, and then integrate my comments with the review changes. :)

**Comment by cticenhour** [comment]
> The failed tests are probably out of the scope of this PR (i.e. one of them uses THM which we're not including here).

@meltawila Could also be related to our MOOSE submodule being ahead of Cardinal's; that would be my guess with only a few tests failing. 

Thanks for checking with your workflow!

File: doc/content/getting_started/installation.md
Comment (cticenhour) [reply doc/content/getting_started/installation.md]: Addressed in 2927d33

File: doc/content/getting_started/installation.md
Comment (cticenhour) [reply doc/content/getting_started/installation.md:90]: Addressed in 2927d33

**Comment by cticenhour** [comment]
With the most-recent documentation commit, a small adjustment to TMAP8 is now required. idaholab/TMAP8#153 is pending review and merge.

File: doc/config.yml
Comment (simopier) [inline doc/config.yml]: Why are you removing ray tracing?

Comment (simopier) [reply doc/config.yml]: Oh sorry, I just saw that it was there twice. Good catch!

File: doc/config.yml
Comment (simopier) [inline doc/config.yml:128]: Why is Cardinal commented out?

File: doc/sqa_fenix.yml
Comment (simopier) [inline doc/sqa_fenix.yml:23]: Again, I'm wondering why cardinal is commented out.

**Review by simopier — CHANGES_REQUESTED** [review]
Just a couple questions.

File: doc/config.yml
Comment (cticenhour) [reply doc/config.yml:128]: Cardinal doesn't use the MOOSE SQA system.

File: doc/config.yml
Comment (cticenhour) [reply doc/config.yml]: Alphabetical module listing is our friend 😄

Comment (cticenhour) [reply doc/config.yml]: It wasn't there twice - I just moved it.

File: doc/sqa_fenix.yml
Comment (cticenhour) [reply doc/sqa_fenix.yml:23]: Again, Cardinal doesn't use the SQA system. This is there as a placeholder if it does (and a reminder to devs that we didn't neglect it). I say this as a comment on Line 5.

File: doc/config.yml
Comment (cticenhour) [reply doc/config.yml:128]: I also say this in a comment on Line 109

**Comment by cticenhour** [comment]
Based on a discussion today with @loganharbour, I think I will adjust this script slightly and the `Makefile` in order to facilitate the running of the script as part of the normal `make` command. This would bring the build process back to something more in-line with a "normal" MOOSE application and ease potential build complications for users.

**Comment by cticenhour** [comment]
One last adjustment to MOOSE was required, which has now been merged. Once that makes it to the MOOSE submodule, I think we'll be ready to go.

**Comment by GiudGiud** [comment]
I think the desired submodule update was not done yet

File: doc/content/getting_started/installation.md
Comment (cticenhour) [inline doc/content/getting_started/installation.md:67]:
```suggestion
git submodule update --init tmap8
git submodule update --init cardinal
```

File: doc/content/getting_started/installation.md
Comment (cticenhour) [inline doc/content/getting_started/installation.md]:
```suggestion
```

Cardinal is now built within the Makefile rules rather then being done manually.

File: doc/content/getting_started/installation.md
Comment (simopier) [inline doc/content/getting_started/installation.md]:
```suggestion
Then navigate to the FENIX clone directory and download the MOOSE, TMAP8, and Cardinal submodules:
```

**Review by simopier — CHANGES_REQUESTED** [review]
Just a small documentation fix given the latest update.

**Comment by cticenhour** [comment]
The TMAP8 portion of this PR has been superceded by #65, so the TMAP8 components will need to be removed. Differing errors are occurring during the Cardinal-inclusive build on Linux CI and on a local container duplicating the CI environment. I will tackle this when back from England.

**Comment by cticenhour** [comment]
The remaining failure is related to compiler warnings coming out of Cardinal/OpenMC. I intend to fix these up (or set them up to be ignored in the case of OpenMC), but that will require 1 libMesh PR, 1 Cardinal PR, and then a follow-on PR here to update Cardinal. 

So as not to hold this up any further, I am simply turning off `-Werror` checking for the debug PR/devel recipe in idaholab/civet_recipes#1841. I have made an issue (#68) that this is a problem that should be resolved per our code standards and code cleanliness guidelines.

I will work on fixing this up as soon as I am able, but it might have to wait until the next FY.

File: doc/content/getting_started/installation.md
Comment (simopier) [inline doc/content/getting_started/installation.md:146]: I cannot point to it or make a suggestion there directly, but the instructions to update FENIX should also be updated. 
It currently only update moose, and should update TMAP8 and Cardinal with something  

```
git submodule update moose
git submodule update tmap8
git submodule update cardinal
```
Note that this change should occur in two different places.

**Review by simopier — CHANGES_REQUESTED** [review]
Just a minor request, all good after that!

File: doc/content/getting_started/installation.md
Comment (cticenhour) [reply doc/content/getting_started/installation.md:146]: Addressed in ffa88f9.

**Review by simopier — DISMISSED** [review]
Thank you!

**Review by loganharbour — COMMENTED** [review]
You want cardinal to be a _requirement_ of fenix?

**Comment by cticenhour** [comment]
@loganharbour This was the intent yes. If this ends up being onerous or disruptive from a CI perspective, we can always make it optional. Is this what you are wary of?

@simopier do you have a preference based on Logan's question?

**Comment by loganharbour** [comment]
No CI concerns. User concerns. In almost all cases where we have coupling, it's optional. If the application has features that do not depend on the coupling application, it should be able to be optional. 

Take a moose dev who is trying to make a patch. IMO don't make them compile cardinal unless they need to.

**Comment by cticenhour** [comment]
Sounds like we're measuring by degrees here. We very often make folks compile multiple MOOSE-based submodules for an application which are "on" at all times. Cardinal's relative weight is the defining difference, and I don't think it's that onerous since we're only adding OpenMC and DAGMC only - nekRS is explicitly left out. 

That being said, I don't have a strong preference; I just wanted to make sure the build and setup worked smoothly for users and devs. The Makefile and source changes to make it optional are not hard to do. As PI, I want to give @simopier the final say.

**Comment by loganharbour** [comment]
I feel pretty strongly about it. This just isn't how we do it it. Not being able to compile it on its own removes semblance of what product this is. And in the future when we make dynamic loading the path forward, you'll have to separate them. Potentially allowing things to mix in to the code base that are cardinal dependent.

**Review by meltawila — APPROVED** [review]
Should we update `doc/content/getting_started/hpc` as well? 
Mainly for adding instructions to build PETSc, libMesh, and WASP as I think it'd be required now. I used:
```
./moose/scripts/update_and_rebuild_petsc.sh
./moose/scripts/update_and_rebuild_libmesh.sh
./moose/scripts/update_and_rebuild_wasp.sh
```

**Comment by cticenhour** [comment]
@meltawila ~~good call - the HPC instructions need to be updated a bit.~~ EDIT: Actually, we do `include` common directions between files. They are updated automatically with these changes.

However, I don't think we need to add the update and rebuild instructions. They should be provided with the latest `moose-dev` module. Has this not been your experience?

File: Makefile
Comment (simopier) [inline Makefile]: Doesn't this define the default as NOT having CARDINAL? 

I think the default should be to include cardinal.

**Review by simopier — CHANGES_REQUESTED** [review]

File: Makefile
Comment (cticenhour) [reply Makefile]: Addressed in 93cb186.

File: Makefile
Comment (simopier) [inline Makefile:50]:
```suggestion
# ENABLE_CARDINAL = yes by default, but is set to "no" automatically if Cardinal is not found via CARDINAL_DIR. 
ifeq ($(ENABLE_CARDINAL),yes)
```

File: config/check_deps.mk
Comment (simopier) [inline config/check_deps.mk:4]: What is that for?

File: config/check_deps.mk
Comment (simopier) [inline config/check_deps.mk]:
```suggestion
  $(warning $n"Cardinal does not seem to be available. If usage of Cardinal is desired within FENIX, make sure that either the submodule is checked out$nor that CARDINAL_DIR points to a location with the Cardinal source. $n In the meantime, FENIX will be built without Cardinal.")
```

File: unit/Makefile
Comment (simopier) [inline unit/Makefile:40]: Great catch!

File: doc/content/getting_started/installation.md
Comment (simopier) [inline doc/content/getting_started/installation.md:62]: The documentation needs to be updated to have a `Tips` message to explain how they can build FENIX without Cardinal if they desire.

File: doc/content/getting_started/installation.md
Comment (simopier) [inline doc/content/getting_started/installation.md:141]: Just as above, this should be updated to tell users how to not use Cardinal.

**Review by simopier — CHANGES_REQUESTED** [review]
Some small questions and a minor request for a documentation update to explain how to not build with Cardinal.

File: config/check_deps.mk
Comment (cticenhour) [reply config/check_deps.mk:4]: It's for spacing and newlines in the printed output of the warning and info messages later in this document. See the references to `$n` there.

File: doc/content/getting_started/installation.md
Comment (simopier) [inline doc/content/getting_started/installation.md:192]:
```suggestion

!alert! tip title=Disabling Cardinal
Cardinal is an optional dependency of FENIX and can be disabled to speed up build times, if Cardinal
capabilities are not needed. This can be done by simply not downloading the Cardinal submodule
(i.e., not performing the `git submodule update cardinal` command as well as the following Cardinal
dependency update commands above.)
!alert-end!

```

**Review by simopier — CHANGES_REQUESTED** [review]
One last thing

**Review by simopier — APPROVED** [review]
Looks great to me, and this is a great contribution! Thank you @cticenhour!!

**Comment by simopier** [comment]
@loganharbour, do you want to give this a another review before we merge it?

**Review by loganharbour — COMMENTED** [review]
Looks good to me. We just need a no optional recipe

**Comment by cticenhour** [comment]
@loganharbour I had already updated the follow-on recipe PR to include it if you want to do it there, or I can pull it out separately. 

EDIT: We're doing it separately, idaholab/civet_recipes#1874

**Review by loganharbour — APPROVED** [review]
Let's do it

**Pull request merged**
